### PR TITLE
check close before send delay command to hyperstart chan

### PR DESCRIPTION
**Pull request by gnawux**
see http://ci.hypercontainer.io:8080/job/hyperd-auto/330/consoleFull

Signed-off-by: Wang Xu <gnawux@gmail.com>

### CODE REVIEW

**Comment by gnawux** [comment]
however, I am not sure should I consider the chan might block at some case.

File: hyperstart/libhyperstart/json.go [hunk 1]
```diff
@@ -262,6 +262,11 @@ func handleMsgToHyperstart(h *jsonBasedHyperstart, conn io.WriteCloser) {
 					// delay version-awared command
 					h.Log(TRACE, "delay version-awared command :%d", cmd.Code)
 					time.AfterFunc(2*time.Millisecond, func() {
+						h.RLock()
+						defer h.RUnlock()
```
Comment (laijs) [inline hyperstart/libhyperstart/json.go]: defer is called when the exit of the function, not when the exit of the scope

Comment (gnawux) [inline hyperstart/libhyperstart/json.go]: It's an anonymous func....

Comment (laijs) [inline hyperstart/libhyperstart/json.go]: Ok...

Comment (laijs) [inline hyperstart/libhyperstart/json.go]: it seams the lock is also needed in the handleMsgFromHyperstart()

File: hyperstart/libhyperstart/json.go [hunk 2]
```diff
@@ -262,6 +262,11 @@ func handleMsgToHyperstart(h *jsonBasedHyperstart, conn io.WriteCloser) {
...
+						if h.closed {
+							return
```
Comment (laijs) [inline hyperstart/libhyperstart/json.go]: an error is needed to be sent to cmd.result

Comment (gnawux) [inline hyperstart/libhyperstart/json.go]: looks so, will add it.

File: hyperstart/libhyperstart/json.go (hunk 1 above)
Comment (gnawux) [inline hyperstart/libhyperstart/json.go]:
```
			h.ctlChan <- &hyperstartCmd{Code: res.Code, retMsg: res.Message}
```

this one?

**Comment by laijs** [comment]
> however, I am not sure should I consider the chan might block at some case.

it is a possible, and cause deadlock.
it seems defer + recovery() proper.

**Comment by gnawux** [comment]
I don't like the taste of `recover()`.... ok, it's a rare case....

**Comment by laijs** [comment]
one of the ways to avoid the deadlock
```go
diff --git a/hyperstart/libhyperstart/json.go b/hyperstart/libhyperstart/json.go
index 852a690..76ffbf0 100644
--- a/hyperstart/libhyperstart/json.go
+++ b/hyperstart/libhyperstart/json.go
@@ -82,6 +82,13 @@ func (h *jsonBasedHyperstart) Log(level hlog.LogLevel, args ...interface{}) {
 }

 func (h *jsonBasedHyperstart) Close() {
+       go func() {
+               for cmd := range h.ctlChan {
+                       if cmd.Code != hyperstartapi.INIT_ACK && cmd.Code != hyperstartapi.INIT_ERROR {
+                               cmd.result <- fmt.Errorf("hyperstart closed")
+                       }
+               }
+       }()
        h.Lock()
        defer h.Unlock()
        if !h.closed {
@@ -97,11 +104,6 @@ func (h *jsonBasedHyperstart) Close() {
                close(h.ctlChan)
                close(h.streamChan)
                close(h.processAsyncEvents)
-               for cmd := range h.ctlChan {
-                       if cmd.Code != hyperstartapi.INIT_ACK && cmd.Code != hyperstartapi.INIT_ERROR {
-                               cmd.result <- fmt.Errorf("hyperstart closed")
-                       }
-               }
                h.closed = true
        }
 }
```

File: hyperstart/libhyperstart/json.go (hunk 1 above)
Comment (laijs) [inline hyperstart/libhyperstart/json.go]: yes

**Comment by gnawux** [comment]
@laijs I think you could close this and push a new PR

**Comment by gnawux** [comment]
@laijs updated

**Comment by gnawux** [comment]
I am considering should I use recover to avoid a lock in a hot path

```
h.ctlChan <- &hyperstartCmd{Code: res.Code, retMsg: res.Message}
```

**Comment by laijs** [comment]
it seams better to avoid the lock in handleMsgFromHyperstart()

**Comment by gnawux** [comment]
ok, same idea. I will use the ugly recover instead.

**Comment by gnawux** [comment]
updated again, with recover instead of lock

**Pull request merged**